File: web/src/types/song.ts
/**
 * 歌曲相关类型定义
 */

// 歌曲来源平台
export type MusicSource = 'netease' | 'qqmusic' | 'kuwo' | 'kugou' | 'migu' | 'local' | 'database'

// 本地文件详情 (Song.local_files 元素, 对应后端 Song.local_files property)
export interface LocalFile {
    id: number
    source_id?: string
    path: string
    quality: string // 'PQ' | 'HQ' | 'SQ' | ...
    format: string  // 'MP3' | 'FLAC' | 'UNK' | ...
}

// 歌曲基础信息
export interface Song {
    id: number
    title: string
    artist: string
    album: string
    source: MusicSource
    source_id: string
    cover?: string
    duration?: number
    local_path?: string
    is_favorite: boolean
    status: SongStatus
    publish_time?: string
    created_at?: string
    found_at?: string
    played_at?: string // 播放/浏览时间 (历史页)
    available_sources?: string[]
    quality?: string // SQ, HQ, Hi-Res
    quality_details?: string
    local_files?: LocalFile[]
}

// 歌曲状态
export type SongStatus = 'PENDING' | 'PENDING_METADATA' | 'DOWNLOADED' | 'FAILED'

// 搜索结果中的歌曲 (后端 SongInfoResponse: cover_url 而非 cover)
export interface SearchSong {
    id: string
    title: string
    artist: string
    album: string
    source: MusicSource
    cover?: string
    cover_url?: string
    duration?: number
    publish_time?: string
    picId?: string
    // 部分搜索场景 (GDStudio 下载检索) 会附带音质/大小信息
    quality?: number | null
}

// 歌曲列表响应 (统一分页: page + page_size)
export interface SongListResponse {
    items: Song[]
    total: number
    page: number
    page_size: number
    total_pages: number
}

// 下载请求
// 播放器侧下载 (POST /api/download_audio)
export interface DownloadRequest {
    title: string
    artist: string
    album: string
    source: MusicSource
    songId: string
    picUrl?: string
}

// 直接下载请求 (搜索结果 → POST /api/library/download)
export interface DirectDownloadRequest {
    title: string
    artist: string
    album: string
    source: string
    source_id: string
    quality?: number
    cover_url?: string
}

// 下载结果
export interface DownloadResult {
    local_path: string
    quality: number
    has_lyric: boolean
    song_id?: string | number
}

File: app/models/song.py
"""
Song模型 - 歌曲数据模型定义

此文件定义了Song数据模型，用于表示音乐监控系统中的歌曲实体。
采用核心-扩展模式:
1. Song: 聚合逻辑实体 (UI显示用, 优先QQ数据)
2. SongSource: 具体平台的音频元数据 (数据用)

Author: music-monitor development team
Updated: 2026-01-27
"""
from sqlalchemy import Column, Integer, String, DateTime, Boolean, ForeignKey, JSON, UniqueConstraint
from sqlalchemy.orm import relationship
from datetime import datetime
from app.models.base import Base

class Song(Base):
    """
    核心歌曲实体 (Display Entity)
    逻辑主键 (Artist + Title + Album)
    """
    __tablename__ = "songs"

    id = Column(Integer, primary_key=True, index=True)
    unique_key = Column(String, unique=True, index=True, nullable=False) # UUID-based unique key
    artist_id = Column(Integer, ForeignKey("artists.id"), nullable=True)
    
    # Display Information (Prioritized)
    title = Column(String, index=True, nullable=False)
    album = Column(String)
    
    # 优先展示用的封面和时间 (QQ > Netease)
    cover = Column(String, nullable=True)
    publish_time = Column(DateTime, nullable=True)
    
    created_at = Column(DateTime, default=datetime.now)
    
    # Status
    is_favorite = Column(Boolean, default=False)
    status = Column(String, default="PENDING") # PENDING / DOWNLOADED / ERROR
    local_path = Column(String, nullable=True) # If downloaded locally
    last_enrich_at = Column(DateTime, nullable=True) # Last time enrichment was attempted
    
    # Relationships
    artist = relationship("Artist", back_populates="songs")
    sources = relationship("SongSource", back_populates="song", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<Song(title={self.title})>"

    @property
    def localFiles(self):
        """
        Helper for API response: returns list of local file details
        """
        result = []
        if not self.sources:
            return result
        for s in self.sources:
            if s.source == 'local':
                 data = s.data_json or {}
                 result.append({
                     "id": s.id,
                     "source_id": s.source_id,
                     "path": s.url,
                     "quality": data.get('quality', 'PQ'),
                     "format": data.get('format', 'UNK')
                 })
        return result

    @property
    def availableSources(self):
        """
        Helper for API response: returns list of unique sources
        """
        if not self.sources:
            return [self.local_path] if self.local_path else []
        return list(set([s.source for s in self.sources]))

class SongSource(Base):
    """
    歌曲来源映射 (Implementation Entity)
    """
    __tablename__ = "song_sources"
    __table_args__ = (
        UniqueConstraint('song_id', 'source', 'source_id', name='uq_song_source'),
    )

    id = Column(Integer, primary_key=True, index=True)
    song_id = Column(Integer, ForeignKey("songs.id"), nullable=False, index=True)
    
    source = Column(String, nullable=False)    # 'qqmusic', 'netease', 'local'
    source_id = Column(String, nullable=False) # mid / filename
    
    # Source Specific Meta
    cover = Column(String, nullable=True)
    duration = Column(Integer, default=0)
    url = Column(String, nullable=True)  # Playback URL if available
    
    # Flexible Data (Quality, Lyric, etc.)
    data_json = Column(JSON, nullable=True)
    
    song = relationship("Song", back_populates="sources")

    def __repr__(self):
        return f"<SongSource(source={self.source}, id={self.source_id})>"
